plugins/inputs/openstack/openstack.go

```diff
@@ -0,0 +1,1059 @@
+type flavorMap map[string]flavors.Flavor
+
+// subnetMap maps a subnet id to a Subnet struct.
+type subnetMap map[string]subnets.Subnet
+
+// portMap maps a port id to a Port struct.
+type portMap map[string]ports.Port
+
+// networkMap maps a network id to a Network struct.
+type networkMap map[string]networks.Network
+
+// aggregateMap maps a aggregate id to a Aggregate struct.
+type aggregateMap map[int]aggregates.Aggregate
+
+// agentMap maps a agent id to a agents struct.
+type agentMap map[string]agents.Agent
+
+// volume is a structure used to unmarshal raw JSON from the API into.
+type volume struct {
+	volumes.Volume
+	volumetenants.VolumeTenantExt
+}
+
+// volumeMap maps a volume id to a volume struct.
+type volumeMap map[string]volume
+
+// storagePoolMap maps a storage pool name to a StoragePool struct.
+type storagePoolMap map[string]schedulerstats.StoragePool
+
+// OpenStack is the main structure associated with a collection instance.
+type OpenStack struct {
+	// Configuration variables
+	IdentityEndpoint   string
+	Domain             string
+	Project            string
+	Username           string
+	Password           string
+	EnabledServices    []string
+	ServerDiagnotics   bool
+	InsecureSkipVerify bool
+
+	// Locally cached clients
+	identity *gophercloud.ServiceClient
+	compute  *gophercloud.ServiceClient
+	volume   *gophercloud.ServiceClient
+	network  *gophercloud.ServiceClient
+	stack    *gophercloud.ServiceClient
+
+	// Locally cached resources
+	services      serviceMap
+	projects      projectMap
+	hypervisors   hypervisorMap
+	flavors       flavorMap
+	servers       serverMap
+	volumes       volumeMap
+	storagePools  storagePoolMap
+	subnets       subnetMap
+	ports         portMap
+	networks      networkMap
+	aggregates    aggregateMap
+	nova_services nova_serviceMap
+	agents        agentMap
+	stacks        stackMap
+	diag          serverDiag
+}
+
+// Description returns a description string of the input plugin and implements
+// the Input interface.
+func (o *OpenStack) Description() string {
+	return "Collects performance metrics from OpenStack services"
+}
+
+// sampleConfig is a sample configuration file entry.
+var sampleConfig = `
+  ## This is the recommended interval to poll.
+  interval = '30m'
+
+  ## The identity endpoint to authenticate against and get the
+  ## service catalog from
+  identity_endpoint = "https://my.openstack.cloud:5000"
+
+  ## The domain to authenticate against when using a V3
+  ## identity endpoint.  Defaults to 'default'
+  domain = "default"
+
+  ## The project to authenticate as
+  project = "admin"
+
+  ## The user to authenticate as, must have admin rights
+  username = "admin"
+
+  ## The user's password to authenticate with
+  password = "Passw0rd"
+
+  ## Services to be enabled
+  #enabled_services = ["stacks","services", "projects", "hypervisors", "flavors", "servers", "volumes", "storage" , "subnets", "ports", "networks", "aggregates", "nova_services", "agents"]
```

Comment:
Please be nice and spend some test instead of making this look like a commented option...
```suggestion
  ## Available services are: 
  ##  "agents", "aggregates", "flavors", "hypervisors", "networks", "nova_services",
  ##  "ports", "projects", "servers", "services", "stacks", "storage", "subnets", "volumes"
```